Survive Conway fork

ffakenz opened the pull request:
<!-- Describe your change here -->

Required for Conway Support https://github.com/input-output-hk/hydra/issues/1177

🛡️ Adapting Cardano and chain-sync clients to survive post the Conway era fork.

- While observing blocks in Conway mode, we convert the block to Babbage using CBOR encoding and decoding.

- Most Cardano client query calls need prior knowledge of the current CardanoEra. To address this, we added `queryCurrentEra` to fetch the ongoing era from the running node before other Cardano queries.
 
- Modifications were made to accommodate the `queryProtocolParameters` function, requiring adjustments in e2e scenarios and faucet. This ensures querying the current era before obtaining pparams from the node.
 
- In the event of unhandled scenarios, exceptions are thrown.

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ffakenz committed:
Draft e2e

ffakenz committed:
Output stderr on process failure

To improve debugging when withHydraNode exits with code 1.

ch1bo committed:
Convert Conway blocks to Babbage on chain sync

This makes it possible to keep our code base to stay on Babbage, while
we receive and understand Conway blocks.

The current draft is using wrapCBORinCBOR/unwrapCBORinCBOR to make the
decodeShelleyBlock side compile, although this should not be needed.

ffakenz committed:
Draft e2e on restart

ffakenz committed:
Add queryCurrentEra to our cardano client

This allow us to pass the current era to our queries and then
dinamically obtain the cardano era in mode and the shelly base era
from it.

ch1bo committed:
Move construction of QueryInEra based on era into a function

ffakenz committed:
Apply fixes after rebase from offline-mode changes

ffakenz committed:
Fix UTxO.fromApi to downgrade conway TxOuts to babbage

ffakenz committed:
Add cardano era to RunningNode

This avoids boilerplate in cluster specs.

Also fixed queryUTxO and prepare baseline to fix queryProtocolParameters

ffakenz committed:
Fix query for protocol-parameters to coerce any era into Babbage

ffakenz committed:
Remove cardano era from RunningNode type

As it can change over time.

ffakenz committed:
Do not pass era to buildTransaction

Instead query it before querying for pparams.

ffakenz committed:
Minor renaming to use coerce

ffakenz committed:
Remove fork detection e2e specs

they are not needed as now we can survive the fork.

ffakenz committed:
Minor formatting changes

ffakenz committed:
Add missing query exceptions to resolve pending TODO comments

ffakenz committed:
Update CHANGELOG

ffakenz committed:
Make tui compile

ffakenz committed:
Minor haddock improvements

ch1bo commented on hydra-cardano-api/src/Cardano/Api/UTxO.hs:
```diff
@@ -84,8 +86,35 @@ min = UTxO . uncurry Map.singleton . Map.findMin . toMap
 
-fromApi :: Cardano.Api.UTxO Era -> UTxO
-fromApi = coerce
+fromApi :: Cardano.Api.UTxO era -> UTxO
```
This is doing a bit more now and a haddock would be great to have here now.

ch1bo commented on hydra-cluster/src/CardanoClient.hs:
```diff
@@ -108,18 +110,20 @@ waitForUTxO ::
   IO ()
-waitForUTxO networkId nodeSocket utxo =
-  forM_ (snd <$> UTxO.pairs utxo) forEachUTxO
+waitForUTxO networkId nodeSocket utxo = do
```
I like that you avoided the `era` parameter here and do it internal to this function. On the other hand, the `waitForPayment` above was changed - why make it inconsistent?

Should: `queryCurrentEra` in `waitForPayment` as it's more consistent.

ffakenz replied:
nice catch!

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Faucet.hs:
```diff
@@ -48,19 +49,21 @@ data FaucetLog
   Tracer IO FaucetLog ->
   IO UTxO
-seedFromFaucet node@RunningNode{networkId, nodeSocket} receivingVerificationKey lovelace tracer = do
+seedFromFaucet node@RunningNode{networkId, nodeSocket} era receivingVerificationKey lovelace tracer = do
```
We are using `seedFromFaucet` in a *lot* of places.. if we would avoid changing this signature, the diff would be a lot smaller as well.

Could: not change `seedFromFaucet` and `seedFromFaucet_`

ffakenz replied:
nice catch!

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -227,9 +232,10 @@ singlePartyHeadFullLifeCycle ::
+singlePartyHeadFullLifeCycle tracer workDir node hydraScriptsTxId = do
+  AnyCardanoEra era <- queryCurrentEra networkId nodeSocket QueryTip
+  (`finally` returnFundsToFaucet tracer node era Alice) $ do
+    refuelIfNeeded tracer node era Alice 25_000_000
```
Must: not re-use the queried era for too long!

This scenario would fail if the network forks between `queryCurrentEra` and `returnFundsToFaucet` in the end. In general, we should not re-use the era longer than a few seconds IMO to write as robust code as possible.

ffakenz replied:
nice catch!

ch1bo commented on hydra-node/src/Hydra/Chain/CardanoClient.hs:
```diff
@@ -245,69 +274,94 @@ queryEpochNo networkId socket queryPoint = do
+    =<< ( mkQueryInEra era QueryProtocolParameters
+            >>= runQuery networkId socket queryPoint
+            >>= throwOnEraMismatch
+        )
```
Could: avoid parenthesis and append `>>= coercePParamsToLedgerEra`

ffakenz replied:
nice catch!

ch1bo commented on hydra-node/src/Hydra/Chain/CardanoClient.hs:
```diff
@@ -325,15 +379,21 @@ queryUTxOWhole networkId socket queryPoint = do
 -- Throws at least 'QueryException' if query fails.
-queryStakePools :: NetworkId -> SocketPath -> QueryPoint -> IO (Set PoolId)
+queryStakePools ::
+  -- | Filepath to the cardano-node's domain socket
```
Must: fix this haddock position.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/ScriptRegistry.hs:
```diff
@@ -157,12 +160,16 @@ registryUTxO scriptRegistry =
   m ScriptRegistry
-queryScriptRegistry networkId socketPath txId = do
-  utxo <- liftIO $ queryUTxOByTxIn networkId socketPath QueryTip candidates
+queryScriptRegistry networkId socketPath txId era = do
```
Should: keep the signature consistent within the module

The `publishHydraScripts` does not take the `CardanoEra`, but this one does. What's the rationale for not being consistent here? IMO we could just `queryCurrentEra` in here like we do in `publishHydraScripts`

ffakenz replied:
nice catch!

ch1bo commented on hydra-node/src/Hydra/Chain/Direct.hs:
```diff
@@ -308,6 +317,22 @@ chainSyncClient handler wallet startingPoint =
               onRollForward handler header txs
               pure clientStIdle
+            BlockInMode _ block ConwayEraInCardanoMode -> do
+              -- TODO: uses cardano-api:internal
```
What is our plan for this TODO?

ffakenz replied:
From our discussion seems the best plan would be to open a PR upstream to demonstrate the need to expose this data-constructor and start the discussion with the cardano-api team.

Once this thing is expose we can remove this fragile dependency.

ch1bo replied:
Yes. Let's keep a note on the TODO that we should remove this once we have `ShelleyBlock` available on the normal cardano-api library.

ch1bo commented on hydra-node/src/Hydra/Chain/CardanoClient.hs:
```diff
@@ -19,6 +22,10 @@ data QueryException
+  | QueryProtocolParamsEraNotSupported Text
+  | QueryProtocolParamsEncodingFailureOnEra Text Text
+  | QueryEraNotInCardanoModeFailure Text
+  | QueryNotShelleyBasedEraException Text
```
Could: hold proper types in these constructors instead of `Text`

```suggestion
  | QueryNotShelleyBasedEraException (forall era. CardanoEra era)
```

..should work

ffakenz replied:
Used AnyCardanoEra instead for this purpose, as it was easier.

ch1bo commented on hydra-node/src/Hydra/Chain/CardanoClient.hs:
```diff
@@ -36,6 +47,13 @@ instance Exception QueryException where
     QueryEraMismatchException EraMismatch{ledgerEraName, otherEraName} ->
       printf "Connected to cardano-node in unsupported era %s. Please upgrade your hydra-node to era %s." otherEraName ledgerEraName
     QueryProtocolParamsConversionException err -> show err
+    QueryProtocolParamsEraNotSupported unsupportedEraName -> show unsupportedEraName
```
This error explanation is not very good as it only shows `ByronEra`. Maybe you missed providing a good error message here.

ch1bo reviewed (changes requested):
I think we should avoid re-using the queried era too much! It introduces a lot of code changes in our e2e tests, and goes "in the wrong direction": if we realize that querying the era in a dedicated connection (as we do now) is too slow, we would want to compose the `QueryCurrentEra` with any of the other queries on the same connection to the `cardano-node`.. and that would happen _in_ `queryUTxO` et al.

ffakenz committed:
Remove era as arg to some functions to avoid reusing the same era, as it can change over time

also added haddock to UTxO.fromApi and applied somew formatting changes.

ch1bo committed:
Change queryUTxO to queryCurrentEra in the same n2c connection

This allows us to keep the same era agnostic interface, while we convert
the arbitrary era UTxO into our canonical (Babbage) UTxO

ch1bo commented on hydra-node/src/Hydra/Chain/Direct.hs:
```diff
@@ -129,8 +135,10 @@ loadChainContext config party = do
+  CardanoEra era ->
   IO (TinyWallet IO)
-mkTinyWallet tracer config = do
+mkTinyWallet tracer config era = do
```
@ffakenz This is an `era` argument we should be able to get rid as well (it impacts quite some tests)

In detail, this means we would need to make `queryProtocolParameters` also query the current era on the same connection. It might be a bit tricky to witness the right `era` types in there, but that would be a good next step.

ffakenz replied:
nice catch!

locallycompact commented on hydra-cluster/src/Hydra/Cluster/Faucet.hs:
```diff
@@ -1,6 +1,6 @@
 module Hydra.Cluster.Faucet where
 
-import Hydra.Cardano.Api
+import Hydra.Cardano.Api hiding (cardanoEra, queryCurrentEra)
```
Can we switch, in general, to using explicit imports from hydra-cardano-api. `hiding` is in general confusing for provenance and explicit imports while verbose are immediately clear to the reader.

ffakenz committed:
Query era as part of queryProtocolParameters

This reduces the code changes and simplifies the diff for review.

ffakenz committed:
Refactor query exceptions to hold AnyCardanoEra types instead of Text

ch1bo commented on hydra-node/src/Hydra/Chain/CardanoClient.hs:
```diff
@@ -19,6 +22,10 @@ data QueryException
+  | QueryProtocolParamsEraNotSupported AnyCardanoEra
+  | QueryProtocolParamsEncodingFailureOnEra AnyCardanoEra Text
+  | QueryEraNotInCardanoModeFailure AnyCardanoEra
+  | QueryNotShelleyBasedEraException AnyCardanoEra
```
:+1:

ch1bo commented on hydra-node/src/Hydra/Chain/CardanoClient.hs:
```diff
@@ -167,15 +187,16 @@ awaitTransaction ::
   ins = keys (UTxO.toMap $ utxoFromTx tx)
   go = do
-    utxo <- queryUTxOByTxIn networkId socket QueryTip ins
+    utxo <- queryUTxOByTxIn networkId socket QueryTip ins era
```
Should: Match `queryUTxOByTxIn` the signature of `queryUTxO` (not require an `era)

This is inconsistent and by removing the era (query the current era in the same node connection), we can avoid quite some changes on the diff.

ffakenz replied:
nice catch!

ch1bo commented on hydra-node/src/Hydra/Chain/CardanoClient.hs:
```diff
@@ -49,7 +68,8 @@ data CardanoClient = CardanoClient
   CardanoClient
-    { queryUTxOByAddress = queryUTxO networkId nodeSocket QueryTip
+    { queryUTxOByAddress = \addresses -> do
+        queryUTxO networkId nodeSocket QueryTip addresses
```
Could: reduce diff

ffakenz replied:
nice catch!

ch1bo commented on hydra-node/src/Hydra/Chain/CardanoClient.hs:
```diff
@@ -223,6 +244,13 @@ queryEraHistory :: NetworkId -> SocketPath -> QueryPoint -> IO (EraHistory Carda
+-- Throws at least 'QueryException' if query fails.
+queryCurrentEra :: NetworkId -> SocketPath -> QueryPoint -> IO AnyCardanoEra
+queryCurrentEra networkId socket queryPoint =
+  runQuery networkId socket queryPoint $ QueryCurrentEra CardanoModeIsMultiEra
```
Could: remove this after we made all era-specific queries use the `queryCurrentEraExpr` internally.

ch1bo commented on hydra-node/src/Hydra/Chain/CardanoClient.hs:
```diff
@@ -245,69 +273,92 @@ queryEpochNo networkId socket queryPoint = do
+  SocketPath ->
+  QueryPoint ->
+  -- | The current running era we can use to query the node
+  CardanoEra era ->
```
Could: Make this consistent with other era-specific queries and do it in "inside" with `queryCurrentEraExpr`

ch1bo commented on hydra-node/src/Hydra/Chain/CardanoClient.hs:
```diff
@@ -245,69 +273,92 @@ queryEpochNo networkId socket queryPoint = do
-          )
-  runQuery networkId socket queryPoint query >>= throwOnEraMismatch
+  AnyCardanoEra era <- queryCurrentEra networkId socket QueryTip
+  mkQueryInEra era QueryProtocolParameters
```
Could: Query in only one connection to the node by using `queryCurrentEraExpr`

Currently this leads to two connections (line 283 and 285)

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/ScriptRegistry.hs:
```diff
@@ -156,13 +158,15 @@ registryUTxO scriptRegistry =
 -- Can throw at least 'NewScriptRegistryException' on failure.
 queryScriptRegistry ::
-  (MonadIO m, MonadThrow m) =>
+  -- | Current network discriminant
```
Nitpick: "Current" does not make sense here. This will never change for a given network. Also, the network discriminant is strictly speaking `Mainnet | Testnet`, where the the network id is a combination of discriminant + magic number.

ffakenz replied:
Updated the haddock to reflect this.

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -526,57 +500,98 @@ spec = around (showLogsOnFailure "EndToEndSpec") $ do
 
     describe "forking eras" $ do
-      it "does report on unsupported era" $ \tracer -> do
-        withClusterTempDir "unsupported-era" $ \tmpDir -> do
```
Could: keep this test, but make it pending to re-enable when we can fork into the next era (after conway)

ch1bo reviewed (approved):
No must-fix remaining, but several changes which could improve consistency and make the diff of this PR smaller still.

ffakenz committed:
Remove cardano era as arg to queryUTxOByTxIn

This reduces the code changes and the diff quite a lot and keeps it
consistent with the rest of the queries.

ffakenz committed:
Revert "Remove fork detection e2e specs"

This reverts commit 2b802f1a9c0394fec615707fee041e31fcbc673b.

ffakenz committed:
Add a note under TODO comment

To explain the reason for this to stay.

Also improved a NetworkId arg haddock.

ffakenz committed:
Refactor cardano client to replace queryCurrentEra in favor of queryCurrentEraExpr

ffakenz committed:
Minor fix after rebase
